paddle/optimizer/parameter_optimizer.cc

```diff
@@ -0,0 +1,67 @@
+#include "parameter_optimizer.h"
+#include "optimizer_factory.h"
+#include <glog/logging.h>
+
+namespace paddle {
+namespace optimizer {
+
+template <class T>
+ParameterOptimizer<T>* ParameterOptimizer<T>::create(
+    const ::std::string &config_proto) {
+  paddle::OptimizerConfig config;
+  CHECK(config.ParseFromString(config_proto) == 0) << "error : optimizer config";
+  CHECK(config_valid(config) == 0) << "error : invalid optimizer config ";
+  ParameterOptimizer<T> *opt = nullptr;
+  switch (config.optimizer_name()) {
+  case "SGD" : opt = new SGDOptimizer<T>(config); break;
```

Comment:
Good suggestion, I'm agreed that parsing protobuf in optimizer is a bad coding style.   However, I found that the optimizer needs an to pack training state and return it to the caller when we need a checkpoint.  I thought configuration proto need to be sending back to the go server side.  That's why I put the config proto into optimizer.